### PR TITLE
[ppml] Add PPML tutorial

### PR DESCRIPTION
**Pull request by liu-shaojun**
## Description
This submission
* add a tutorial for PPML which includes What is PPML, Why PPML, Getting Started with PPML and Develop Your Application with PPML
* add three videos into tutorial: ppml_intro, ppml_e2e_workflow and simplequery_e2e_demo
* add a PPML CLI user manual in submit_job.md
* organize examples into examples.md
* update kms-util readme

### CODE REVIEW

**Comment by jason-dai** [comment]
At the beginning of the tutorial (before the table of contents), we need a very short (one or two sentences) description that talks about what BigDL PPML is from the user point of view (e.g., what types of use cases/applications can be supported).

**Comment by liu-shaojun** [comment]
> At the beginning of the tutorial (before the table of contents), we need a very short (one or two sentences) description that talks about what BigDL PPML is from the user point of view (e.g., what types of use cases/applications can be supported).

Thanks for your comment, the short description is added in the beginning part.

File: ppml/docs/prepare_environment.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        -e RUNTIME_SPARK_MASTER=$K8S_MASTER \
+        -e RUNTIME_K8S_SERVICE_ACCOUNT=spark \
+        -e RUNTIME_K8S_SPARK_IMAGE=$DOCKER_IMAGE \
+        -e RUNTIME_DRIVER_HOST=$LOCAL_IP \
+        -e RUNTIME_DRIVER_PORT=54321 \
+        -e RUNTIME_DRIVER_CORES=1 \
+        -e RUNTIME_EXECUTOR_INSTANCES=1 \
+        -e RUNTIME_EXECUTOR_CORES=8 \
+        -e RUNTIME_EXECUTOR_MEMORY=1g \
+        -e RUNTIME_TOTAL_EXECUTOR_CORES=4 \
```
Comment (glorysdj) [inline ppml/docs/prepare_environment.md:142]: this line need to be removed. we use RUNTIME_EXECUTOR_INSTANCES and RUNTIME_EXECUTOR_CORES to determine RUNTIME_TOTAL_EXECUTOR_CORES

Comment (liu-shaojun) [reply ppml/docs/prepare_environment.md]: thx! removed.

File: ppml/README.md [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;[Step 4. Decrypt and Read Result](#step-4-decrypt-and-read-result) \
+&ensp;&ensp;[3.3 More BigDL PPML Examples](#33-more-bigdl-ppml-examples) \
+[4. Develop your own Big Data & AI applications with BigDL PPML](#4-develop-your-own-big-data--ai-applications-with-bigdl-ppml)
+
+
+## 1. What is BigDL PPML?
+
+Protecting data privacy and confidentiality is critical in a world where data is everywhere. In recent years, more and more countries have enacted data privacy legislation or are expected to pass comprehensive legislation to protect data privacy, the importance of privacy and data protection is increasingly recognized.
+
+To better protect sensitive data, it’s helpful to think about it in all dimensions of data lifecycle: data at rest, data in transit, and data in use. Data being transferred on a network is “in transit”, data in storage is “at rest”, and data being processed is “in use”.
```
Comment (qiyuangong) [inline ppml/README.md:22]: `it’s helpful to think about it in all dimensions of data` changes to `it's necessary to ensure security for all xxxx`

File: ppml/README.md [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+<p align="center">
+  <img src="https://user-images.githubusercontent.com/61072813/177720405-60297d62-d186-4633-8b5f-ff4876cc96d6.png" alt="data lifecycle" width='390px' height='260px'/>
+</p>
+
+For protecting data in transit, enterprises often choose to encrypt sensitive data prior to moving or use encrypted connections (HTTPS, SSL, TLS, FTPS, etc) to protect the contents of data in transit. For protecting data at rest, enterprises can simply encrypt sensitive files prior to storing them or choose to encrypt the storage drive itself. However, the third state, data in use has always been a weakly protected target. There are three emerging solutions seek to reduce the data-in-use attack surface: homomorphic encryption, multi-party computation, and confidential computing.
```
Comment (qiyuangong) [inline ppml/README.md:28]: For protecting changes to To protect

File: ppml/README.md [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+Among these, [Confidential computing](https://www.intel.com/content/www/us/en/security/confidential-computing.html) protects data in use by performing computation in a hardware-based [Trusted Execution Environment (TEE)](https://en.wikipedia.org/wiki/Trusted_execution_environment). [Intel® SGX](https://www.intel.com/content/www/us/en/developer/tools/software-guard-extensions/overview.html) is Intel’s Trusted Execution Environment (TEE), offering hardware-based memory encryption that isolates specific application code and data in memory. [Intel® TDX](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-trust-domain-extensions.html) is the next generation Intel’s Trusted Execution Environment (TEE), introducing new, architectural elements to help deploy hardware-isolated, virtual machines (VMs) called trust domains (TDs).
```
Comment (qiyuangong) [inline ppml/README.md:30]: Among these security technologies, [Confidential Computing]xxx

File: ppml/README.md (hunk 2 above)
Comment (qiyuangong) [inline ppml/README.md:22]: replace " with `.

File: ppml/README.md (hunk 4 above)
Comment (qiyuangong) [inline ppml/README.md:30]: Intel's is using Chinese '

File: ppml/README.md [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+## 2. Why BigDL PPML?
+PPML allows organizations to explore powerful AI techniques while working to minimize the security risks associated with handling large amounts of sensitive data. PPML protects data at rest, in transit and in use: compute and memory protected by SGX Enclaves, storage (e.g., data and model) protected by encryption, network communication protected by remote attestation and Transport Layer Security (TLS), and optional Federated Learning support. 
+
+<p align="left">
+  <img src="https://user-images.githubusercontent.com/61072813/177922914-f670111c-e174-40d2-b95a-aafe92485024.png" alt="data lifecycle" width='600px' />
+</p>
+
+With BigDL PPML, you can run trusted Big Data & AI applications
+- **Trusted Spark SQL & Dataframe**: with the trusted Big Data analytics and ML/DL support, users can run standard Spark data analysis (such as Spark SQL, Dataframe, MLlib, etc.) in a secure and trusted fashion.
+- **Trusted ML**: with the trusted Big Data analytics and ML/DL support, users can run distributed machine learning (such as MLlib, XGBoost) in a secure and trusted fashion.
```
Comment (qiyuangong) [inline ppml/README.md:43]: ML (Machine Learning)

File: ppml/README.md [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+- **Trusted DL**: with the trusted Big Data analytics and ML/DL support, users can run distributed deep learning (such as BigDL, Orca, Nano, DLlib) in a secure and trusted fashion.
```
Comment (qiyuangong) [inline ppml/README.md:44]: DL (Deep Learning)

File: ppml/README.md [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+- **Trusted FL (Federated Learning)**: TODO
```
Comment (qiyuangong) [inline ppml/README.md:45]: **Trusted FL (Federated Learning)**: with PSI (Private Set Intersection), Secured Aggregation and trusted federated learning support , users can build united model across different parties without compromising privacy, even if these parities have different datasets or features.

**Review by qiyuangong — APPROVED** [review]
LGTM

File: ppml/README.md [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+  ```
+
+### 3.3 More BigDL PPML Examples
+In addition to the above Spark Pi and Python HelloWorld programs running locally, and simplequery application running on the k8s cluster, we also provide other examples including Trusted Data Analysis, Trusted ML, Trusted DL and Trusted FL. You can find these examples in [more examples](./docs/examples.md). 
+
+## 4. Develop your own Big Data & AI applications with BigDL PPML
+
+We provide both Scala and Python APIs to help you develop your own applications with BigDL PPML. 
+
+First thing you need to do is create a `PPMLContext`, who wraps a `SparkSession` and provides read functions to read encrypted data files to plain-text RDD or DataFrame, also provides write functions to save DataFrame to encrypted data files.
```
Comment (liu-shaojun) [inline ppml/README.md:273]: First you need to create a `PPMLContext`, which wraps `SparkSession` and provides methods to read encrypted data file into plain-text RDD/DataFrame and write DataFrame to encrypted data file. Then you can read & write data through `PPMLContext`.

If you are familiar with Spark, you may find that the usage of `PPMLConext` is very similar to Spark.

File: ppml/README.md [hunk 9]
```diff
@@ -0,0 +1,505 @@
...
+
+## 4. Develop your own Big Data & AI applications with BigDL PPML
+
+We provide both Scala and Python APIs to help you develop your own applications with BigDL PPML. 
+
+First thing you need to do is create a `PPMLContext`, who wraps a `SparkSession` and provides read functions to read encrypted data files to plain-text RDD or DataFrame, also provides write functions to save DataFrame to encrypted data files. 
+
+If you are familiar with Spark, then you will find that using `PPMLConext` is very similar to Spark.
+
+### 4.1 PPMLContext In Scala
```
Comment (liu-shaojun) [inline ppml/README.md:277]: ### 4.1 Develop App In Scala

File: ppml/README.md [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+// read from a plain csv file and return a RDD
+val plainCsvPath = "/plain/csv/path"
+val rdd1 = sc.textfile(plainCsvPath) // the default cryptoMode is PLAIN_TEXT
+
+// read from a encrypted csv file and return a RDD
+val encryptedCsvPath = "/encrypted/csv/path"
+val rdd2 = sc.textfile(path=encryptedCsvPath, cryptoMode=AES_CBC_PKCS5PADDING)
+```
+
+### 4.2 PPMLContext In Python
```
Comment (liu-shaojun) [inline ppml/README.md:503]: ### 4.2 Develop App In Python

File: ppml/README.md [hunk 11]
```diff
@@ -0,0 +1,505 @@
...
+
+We also provide `PPMLContext` Python API. Please refer to [this](https://github.com/intel-analytics/BigDL/blob/main/python/ppml/src/bigdl/ppml/README.md)
```
Comment (liu-shaojun) [inline ppml/README.md:505]: To use `PPMLContext` Python API, please refer to [PPMLContext Python API](https://github.com/intel-analytics/BigDL/blob/main/python/ppml/src/bigdl/ppml/README.md).

File: ppml/README.md [hunk 12]
```diff
@@ -0,0 +1,505 @@
...
+       "spark.bigdl.kms.key.primary" -> /your/primary/key/path,
+       "spark.bigdl.kms.key.data" -> /your/data/key/path
+   )
+   
+   val conf: SparkConf = new SparkConf().setMaster("local[4]")
+   
+   val sc = PPMLContext.initPPMLContext(conf, "MyApp", ppmlArgs)
+   ```
+
+#### 4.1.2 Read & Write Files
```
Comment (liu-shaojun) [inline ppml/README.md:376]: #### 2. Read & Write Files

File: ppml/README.md [hunk 13]
```diff
@@ -0,0 +1,505 @@
...
+
+We provide both Scala and Python APIs to help you develop your own applications with BigDL PPML. 
+
+First thing you need to do is create a `PPMLContext`, who wraps a `SparkSession` and provides read functions to read encrypted data files to plain-text RDD or DataFrame, also provides write functions to save DataFrame to encrypted data files. 
+
+If you are familiar with Spark, then you will find that using `PPMLConext` is very similar to Spark.
+
+### 4.1 PPMLContext In Scala
+
+#### 4.1.1 create a PPMLContext
```
Comment (liu-shaojun) [inline ppml/README.md:279]: #### 1. create PPMLContext

File: ppml/README.md [hunk 14]
```diff
@@ -0,0 +1,505 @@
...
+
+1. `create a PPMLContext with appName`
```
Comment (liu-shaojun) [inline ppml/README.md:281]: 1. create PPMLContext with `appName`

File: ppml/README.md [hunk 15]
```diff
@@ -0,0 +1,505 @@
...
+
+   ```scala
+   import com.intel.analytics.bigdl.ppml.PPMLContext
+   
+   val sc = PPMLContext.initPPMLContext("MyApp")
+   ```
+
+   If you want to read/write encrypted files, then you need to provide more information.
+
+2. `create a PPMLContext with appName & ppmlArgs`
```
Comment (liu-shaojun) [inline ppml/README.md:295]: 2. create PPMLContext with `appName` and `ppmlArgs`

File: ppml/README.md [hunk 16]
```diff
@@ -0,0 +1,505 @@
...
+       "spark.bigdl.kms.azure.vault" -> key_vault_name,
+       "spark.bigdl.kms.azure.clientId" -> client_id,
+       "spark.bigdl.kms.key.primary" -> /your/primary/key/path,
+       "spark.bigdl.kms.key.data" -> /your/data/key/path
+   )
+   
+   val sc = PPMLContext.initPPMLContext("MyApp", ppmlArgs)
+   ```
+
+3. `create a PPMLContext with sparkConf & appName & ppmlArgs`
```
Comment (liu-shaojun) [inline ppml/README.md:355]: 3. create PPMLContext with `appName` & `ppmlArgs`  & `sparkConf`

File: ppml/README.md [hunk 17]
```diff
@@ -0,0 +1,505 @@
...
+
+2. `create a PPMLContext with appName & ppmlArgs`
+
+   `ppmlArgs` is a Map, what parameters in the Map depends on what kind of `KeyManagementService(KMS)` you use. we support 3 types of KMS ——`SimpleKeyManagementService`, `EHSMKeyManagementService`, `AzureKeyManagementService`. 
+
+   We use KMS to generate a `primaryKey` and `dataKey`, then we use the keys to encrypt/decrypt data.
+
+   > How to use KMS to generate primaryKey and dataKey, please refer to [this](https://github.com/intel-analytics/BigDL/blob/main/ppml/services/kms-utils/docker/README.md)
+
+   After preparing the keys, now you can create a `PPMLContext` with the following parameters.
```
Comment (liu-shaojun) [inline ppml/README.md:303]: `ppmlArgs` is ppml arguments in a Map, `ppmlArgs` varies according to the kind of Key Management Service (KMS) you are using. Key Management Service (KMS) is used to generate `primaryKey` and `dataKey` to encrypt/decrypt data. We provide 3 types of KMS ——SimpleKeyManagementService, EHSMKeyManagementService, AzureKeyManagementService. 

   Refer to [KMS Utils](https://github.com/intel-analytics/BigDL/blob/main/ppml/services/kms-utils/docker/README.md) to use KMS to generate `primaryKey` and `dataKey`, then you are ready to create **PPMLContext** with `ppmlArgs`.

File: ppml/README.md [hunk 18]
```diff
@@ -0,0 +1,505 @@
...
+   ```
+
+#### 4.1.2 Read & Write Files
+
+After you create a `PPMLContext`, you can use the method in `PPMLContext` to do read/write operations. you can read from a plain file or encrypted file, or write a DataFrame as plain or encrypted file, so you need to specify the `CryptoMode`:
+
+- `plain_text`: no encryption
+- `AES/CBC/PKCS5Padding`: for CSV, JSON and text file
+- `AES_GCM_V1`: for PARQUET only
+- `AES_GCM_CTR_V1`: for PARQUET only
```
Comment (liu-shaojun) [inline ppml/README.md:383]: To read/write data, you should set the `CryptoMode`:
- `plain_text`: no encryption
- `AES/CBC/PKCS5Padding`: for CSV, JSON and text file
- `AES_GCM_V1`: for PARQUET only
- `AES_GCM_CTR_V1`: for PARQUET only

To write data, you should set the `write` mode:
- `overwrite`: Overwrite existing data with the content of dataframe.
- `append`: Append content of the dataframe to existing data or table.
- `ignore`: Ignore current write operation if data / table already exists without any error.
- `error`: Throw an exception if data or table already exists.
- `errorifexists`: Throw an exception if data or table already exists.

<details><summary>expand to see the example of CryptoMode and write mode</summary>
code here
</details>

File: ppml/README.md [hunk 19]
```diff
@@ -0,0 +1,505 @@
...
+&ensp;&ensp;[3.1 BigDL PPML Hello World](#31-bigdl-ppml-hello-world) \
+&ensp;&ensp;[3.2 BigDL PPML End-to-End Workflow](#32-bigdl-ppml-end-to-end-workflow) \
+&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;[Step 0. Preparation your environment](#step-0-preparation-your-environment): detailed steps in [Prepare Environment](https://github.com/liu-shaojun/BigDL/blob/ppml_doc/ppml/docs/prepare_environment.md) \
+&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;[Step 1. Encrypt and Upload Data](#step-1-encrypt-and-upload-data) \
+&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;[Step 2. Build Big Data & AI applications](#step-2-build-big-data--ai-applications) \
+&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;[Step 3. Submit Job](#step-3-submit-job): 4 deploy modes and 2 options to submit job  \
+&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;[Step 4. Decrypt and Read Result](#step-4-decrypt-and-read-result) \
+&ensp;&ensp;[3.3 More BigDL PPML Examples](#33-more-bigdl-ppml-examples) \
+[4. Develop your own Big Data & AI applications with BigDL PPML](#4-develop-your-own-big-data--ai-applications-with-bigdl-ppml)
+
```
Comment (liu-shaojun) [inline ppml/README.md:16]: add **4.1 Develop App in Scala** and **4.2 Develop App in Python** into Table of Contents @PatrickkZ

File: ppml/README.md [hunk 20]
```diff
@@ -0,0 +1,517 @@
...
+	...
+
+// write data
+sc.write(dataFrame = df, cryptoMode = AES_CBC_PKCS5PADDING)
+.mode("overwrite")
+...
+```
+
+</details>
+
```
Comment (liu-shaojun) [inline ppml/README.md:406]: <details><summary>expand to see the examples of reading/writing CSV, PARQUET, JSON and text file</summary>

In the following examples `sc` represents an initialized `PPMLContext`.

read/write CSV file
```
code
```

read/write PARQUET file
```
code
```

...

File: ppml/README.md [hunk 21]
```diff
@@ -0,0 +1,517 @@
...
+The following examples use `sc` to represent a initialized `PPMLContext`
+
+##### CSV
+
+Code Example
```
Comment (liu-shaojun) [inline ppml/README.md:411]: remove 'Code Example', also remove the following 'Code Example' words

File: ppml/README.md [hunk 22]
```diff
@@ -0,0 +1,517 @@
...
+
+First you need to create a `PPMLContext`, which wraps `SparkSession` and provides methods to read encrypted data file into plain-text RDD/DataFrame and write DataFrame to encrypted data file. Then you can read & write data through `PPMLContext`.
+
+If you are familiar with Spark, you may find that the usage of `PPMLConext` is very similar to Spark.
+
+### 4.1 Develop App In Scala
+
+#### 1. create a PPMLContext
+
+1. create a PPMLContext with `appName`
```
Comment (liu-shaojun) [inline ppml/README.md:282]: update 1.2.3 to a.b.c
```
update
1. create a PPMLContext with `appName` 
to 
a. create a PPMLContext with `appName`
```

**Review by glorysdj — APPROVED** [review]
LGTM

**Pull request merged**